language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java

```diff
@@ -183,4 +201,121 @@ public static CodeActionNodeType topLevelNodeInLine(LSContext context, TextDocum
+                Position typeEndCodeBodyPosition = new Position(typeLineRange.endLine().line(),
+                        typeLineRange.endLine().offset());
+                return new NodePositions(startCodePosition, startCodeBodyPosition, typeEndCodeBodyPosition);
+        }
+        return new NodePositions(startCodePosition, null, null);
+    }
+
+    /**
+     * Returns the number of API documentation lines.
+     *
+     * @param member Node for which the doc lines are calculated
+     * @return Number of documentation lines
+     */
+    private static int getNumberOfDocLines(NonTerminalNode member) {
+        if (member.children().size() > 0 && member.children().get(0) instanceof MetadataNode) {
+            LineRange docLineRange = member.children().get(0).lineRange();
+            return docLineRange.endLine().line() - docLineRange.startLine().line() + 1;
+        }
+        return 0;
+    }
+
+    /**
+     * Checks if a give position is within a given range.
+     *
+     * @param start Start position of the range
+     * @param end End position of the range
+     * @param position Cursor position
+     * @param isEndInclusive If the range is included with end position or not
+     * @return If the position is within the range or not
+     */
+    private static boolean isWithinRange(Position start, Position end, Position position, boolean isEndInclusive) {
+        if (start == null || end == null || position == null) {
+            return false;
+        }
+
+        if (position.getLine() > start.getLine() && position.getLine() < end.getLine()) {
+            return true;
+        } else if (position.getLine() == start.getLine()) {
+            return position.getCharacter() >= start.getCharacter() && (position.getLine() < end.getLine()
+                    || (position.getLine() == end.getLine() && (position.getCharacter() < end.getCharacter()
+                    || isEndInclusive && position.getCharacter() == end.getCharacter())));
+        } else if (position.getLine() == end.getLine()) {
+            return (position.getCharacter() < end.getCharacter() || (isEndInclusive
+                    && position.getCharacter() == end.getCharacter())) && (position.getLine() > start.getLine()
+                    || (position.getLine() == start.getLine() && position.getCharacter() >= start.getCharacter()));
+        }
+        return false;
+    }
+
+    private static NodePositions createNodePositions(LineRange lineRange, Position startCodePosition) {
+        Position startCodeBodyPosition = new Position(lineRange.startLine().line(), lineRange.startLine().offset());
+        Position endCodeBodyPosition = new Position(lineRange.endLine().line(), lineRange.endLine().offset());
+        return new NodePositions(startCodePosition, startCodeBodyPosition, endCodeBodyPosition);
+    }
+
+    /**
+     * Inner class to hold on to the positions of a node.
+     */
+    private static class NodePositions {
```

Comment:
Do we need this? Isn't it possible to do the calculations on the fly?

Reply:
At least two of these positions are required for each node position related decision. And if we are to calculate these values separately we have to repeat some operations, such as calculating the number of doc lines, iterating children to get the body start node, etc., where we could calculate all within a single execution if we save them in an object.

Reply:
This was removed with offset based calculations.